### DISCUSSION TITLE
Build isolation

### DISCUSSION THREAD

**Discussion by sr-murthy** [pdm-project/pdm#2281] [General]
Can someone please explain the exact difference in behaviour of `pdm build` between using the `--no-isolation option`, and leaving it out? I've tried to build wheels in my project (using PDM 2.9.2) in a pre-existing Python 3.9.6 virtual env. (created using `virtualenv`) using both options and the behaviour seems to be the same. Confusingly, the CLI output seems to indicate it is trying to create an isolated build env. in both cases:

    $ pdm build -v --no-sdist -d build
    Building wheel...
    pdm.termui: Preparing isolated env for PEP 517 build...
    ...

and

    $ pdm build -v --no-isolation --no-sdist -d build
    Building wheel...
    pdm.termui: Preparing isolated env for PEP 517 build...
    ...

Is this simply a CLI output inconsistency, or is something else going on behind the scenes?

**Comment by frostming** [comment]
> Is this simply a CLI output inconsistency

That's the case

**Reply by sr-murthy** [reply]
@frostming OK, but with `--no-isolation`, which I assume means it does the build in the current virtual env., which already has the build backend dependencies:`importlib-metadata`, `pdm-backend`, `zipp` etc., I see that it is trying to install these:

    ...
    pdm.termui: Installing importlib-metadata@6.8.0...
    ...
    pdm.termui: Installing pdm-backend@2.1.6...
    ...
    pdm.termui: Installing zipp@3.17.0...
    ...

If they already exist in the current virtual env., and we are using `--no-isolation`, why does it need to install them again?

**Reply by sr-murthy** [reply]
I think these build backend dependencies existed in the virtual env. prior to running the command above:

    (venv) $ pdm show pdm-backend
    Name:                  pdm-backend                                                           
    Latest version:        2.1.6
    ...

    (venv) $ pdm show importlib-metadata
    Name:                  importlib-metadata                          
    Latest version:        6.8.0
    ...

    (venv) $ pdm show zipp
    Name:                  zipp                                                       
    Latest version:        3.17.0
    ...

I'm just trying to understand the actual difference in build environments and behaviour with and without `--no-isolation`.

**Reply by frostming** [reply]
> I think these build backend dependencies existed in the virtual env.

It can't be proven by the information given. `pdm show` will show the PyPI version if not installed. The package is installed only if `Installed version` exists in the output

**Reply by sr-murthy** [reply]
I'll update here after retesting this is a new virtual env.
